// @fs/zion-tree-pedigree/src/utils/translationKeyMaps.js
export const ancestorsKeyMap = {
  1: 'tw.pedigree.ancestors.1',
  2: 'tw.pedigree.ancestors.2',
}

export const descendantsKeyMap = {
  1: 'tw.pedigree.descendants.1',
  2: 'tw.pedigree.descendants.2',
}

export const childrenButtonKeyMap = {
  1: 'tw.pedigree.children-button.1',
  2: 'tw.pedigree.children-button.2',
}

export const addChildButtonKeyMap = {
  1: 'tw.pedigree.add-child-button.1',
  2: 'tw.pedigree.add-child-button.2',
}

export const collapseChildrenKeyMap = {
  1: 'tw.pedigree.collapse-children.1',
  2: 'tw.pedigree.collapse-children.2',
}

export const expandChildrenKeyMap = {
  1: 'tw.pedigree.expand-children.1',
  2: 'tw.pedigree.expand-children.2',
}

export const pedigreeViewKeyMap = {
  landscape: 'tw.pedigree.landscape',
  portrait: 'tw.pedigree.portrait',
  sibling: 'tw.pedigree.sibling',
  fanchart: 'tw.pedigree.fanchart',
  descendancy: 'tw.pedigree.descendancy',
}

export const pedigreeHeadingKeyMap = {
  1: 'tw.pedigree.pedigree-heading.1',
  2: 'tw.pedigree.pedigree-heading.2',
}

export const relationshipTypeKeyMap = {
  marriage: 'tw.pedigree.marriage',
  divorce: 'tw.pedigree.divorce',
}

export const collapseParentsKeyMap = {
  1: 'tw.pedigree.collapse-parents.1',
  2: 'tw.pedigree.collapse-parents.2',
}

export const expandParentsKeyMap = {
  1: 'tw.pedigree.expand-parents.1',
  2: 'tw.pedigree.expand-parents.2',
}
