// @fs/zion-tree-pedigree/src/components/shared/PrintButton.tsx
import React from 'react'
import { ContentPrint } from '@fs/zion-icon'
import { useTranslation } from 'react-i18next'
import { i18n } from '@fs/zion-locale'
import { Button } from '@fs/zion-ui'
import { usePreference } from '@fs/zion-preferences'
import { usePedigreeViewOptions } from '../../pedigree-view-options/index'

const fanchartViewMap = {
  birthPlace: 'BIRTH_COUNTRY',
  sources: 'SOURCES',
  stories: 'STORIES',
  photos: 'PHOTOS',
  researchHelps: 'RESEARCH_HELPS',
  ordinances: 'ORDINANCES',
  familyLines: 'FAMILY_LINES',
}

const pathMap = {
  portrait: (pofId) => `/service/tree/tree-data/pdf/portrait-pedigree/${pofId}/spouse/DEFAULT`,
  landscape: (pofId) => `/service/tree/tree-data/pdf/pedigree-chart/${pofId}/spouse/DEFAULT`,
  fanchart: (pofId) => `/service/tree/tree-data/pdf/fan-chart/${pofId}`,
}

type ParamsProps = {
  locale: string
  showOrdinances?: string
  showNamesInRomanScript?: string
  view?: string
  numGenerations?: string
}
type Props = {
  view: 'portrait' | 'landscape' | 'fanchart'
  pofId: string
}
export default function PrintButton({ view, pofId }: Props): JSX.Element {
  const { t } = useTranslation()
  const {
    showTemple,
    prefs: { showTempleOpportunities, fanchartMode, fanchartGens },
  } = usePedigreeViewOptions()
  const showNameInRomanScriptPreference = usePreference('tree.showNamesInRomanScript')

  const gensNumber = Number(fanchartGens || 4) - 1
  const paramsProps: ParamsProps = {
    locale: i18n.language || 'en',
  }
  if (view === 'fanchart') {
    paramsProps.view = fanchartViewMap[fanchartMode || 'familyLines']
    paramsProps.numGenerations = gensNumber?.toString()
  } else {
    paramsProps.showOrdinances = Boolean(showTemple && showTempleOpportunities).toString()
    paramsProps.showNamesInRomanScript = Boolean(showNameInRomanScriptPreference).toString()
  }
  const params = new URLSearchParams(paramsProps)

  const path = pathMap[view]?.(pofId) || ''
  const printHref = `${path}?${params.toString()}`

  return (
    <Button Icon={ContentPrint} to={printHref} emphasis="medium" external target="_blank" data-testid="print-pedigree">
      {t('tw.pedigree.print')}
    </Button>
  )
}
