// @fs/zion-ui/src/tooltip/ComboTooltip.js
import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { deprecatedPropType } from '@fs/zion-frontend-friends'
import DescriptionTooltip, { defaultOpenTimeout } from './DescriptionTooltip'

export default function ComboTooltip(props) {
  const [opening, setOpening] = useState()
  const [expanded, setExpanded] = useState()
  const {
    description,
    linkProps, // Deprecated, remove in v11
    keyboardShortcut,
    openTimeout = defaultOpenTimeout, // [HIDDEN]: For internal Zion-UI use only
  } = props

  useEffect(() => {
    if (opening) {
      const timer = setTimeout(function openDescription() {
        setExpanded(true)
      }, openTimeout)
      return () => {
        clearTimeout(timer)
      }
    }
    return () => {} // return a no-op for eslint consistent return
  }, [opening, openTimeout])

  const handlePopperOpen = () => {
    setOpening(true)
  }

  return (
    <DescriptionTooltip
      {...props}
      description={expanded ? description : undefined}
      keyboardShortcut={expanded ? keyboardShortcut : undefined}
      linkProps={expanded ? linkProps : undefined}
      onOpenFinished={handlePopperOpen}
      openTimeout={0}
    />
  )
}

ComboTooltip.propTypes = {
  /** The label of the tooltip, shown immediately on hover/focus. */
  label: PropTypes.string.isRequired,

  /** Max 140 characters. Shown after 2 seconds of hovering/focusing. */
  description: PropTypes.string.isRequired,

  // Used in Tooltip
  /** A single dom node or component that renders a single dom node to which the tooltip gets attached. */
  target: PropTypes.oneOfType([PropTypes.element, PropTypes.node]).isRequired,

  /** Keyboard shortcut for some action that the anchoring button does. */
  keyboardShortcut: PropTypes.string,

  /** Deprecated: tooltips should not be interactive. */
  linkProps: deprecatedPropType(
    'ComboTooltip: `linkProps` prop has been deprecated. See https://beta.familysearch.org/frontier/zion/?path=/docs/ui-foundation-feedback-overlays-tooltipoverlay'
  ),
}
