// @braze/web-sdk/src/triggers/models/trigger-condition.js
import ts from "./custom-event-data.js";
import rs from "./custom-event-property-data.js";
import is from "./filter-set.js";
import ni from "./in-app-message-click-data.js";
import ns from "./purchase-data.js";
import hs from "./purchase-property-data.js";
import ur from "./push-click-data.js";
import et from "./trigger-events.js";
import { logger as N } from "../../../shared-lib/index.js";
import { getErrorMessage as si } from "../../util/error-utils.js";
export default class ri {
  constructor(e, t) {
    (this.type = e), (this.data = t), (this.type = e), (this.data = t);
  }
  _c(e, t) {
    return ri.Ec[this.type] === e && (null == this.data || this.data.yc(t));
  }
  static sf(e, t) {
    let r = null;
    try {
      r = window.atob(e);
    } catch (t) {
      return N.info("Failed to unencode analytics id " + e + ": " + si(t)), !1;
    }
    return t === r.split("_")[0];
  }
  static fromJson(e) {
    const t = e.type;
    let r = null;
    switch (t) {
      case ri.Xr.OPEN:
      case ri.Xr.Bs:
        break;
      case ri.Xr.Rr:
        r = ns.fromJson(e.data);
        break;
      case ri.Xr.jc:
        r = hs.fromJson(e.data);
        break;
      case ri.Xr.hr:
        r = ur.fromJson(e.data);
        break;
      case ri.Xr.Ee:
        r = ts.fromJson(e.data);
        break;
      case ri.Xr.Cc:
        r = rs.fromJson(e.data);
        break;
      case ri.Xr.Bo:
        r = ni.fromJson(e.data);
    }
    return new ri(t, r);
  }
  bt() {
    return { t: this.type, d: this.data ? this.data.bt() : null };
  }
  static En(e) {
    let t,
      r = null;
    switch (e.t) {
      case ri.Xr.OPEN:
      case ri.Xr.Bs:
        break;
      case ri.Xr.Rr:
        r = new ns(e.d);
        break;
      case ri.Xr.jc:
        (t = e.d || {}), (r = new hs(t.id, is.En(t.pf || [])));
        break;
      case ri.Xr.hr:
        r = new ur(e.d);
        break;
      case ri.Xr.Ee:
        r = new ts(e.d);
        break;
      case ri.Xr.Cc:
        (t = e.d || {}), (r = new rs(t.e, is.En(t.pf || [])));
        break;
      case ri.Xr.Bo:
        r = new ni(e.d);
    }
    return new ri(e.t, r);
  }
}
(ri.Xr = {
  OPEN: "open",
  Rr: "purchase",
  jc: "purchase_property",
  hr: "push_click",
  Ee: "custom_event",
  Cc: "custom_event_property",
  Bo: "iam_click",
  Bs: "test",
}),
  (ri.Ec = {}),
  (ri.Ec[ri.Xr.OPEN] = et.OPEN),
  (ri.Ec[ri.Xr.Rr] = et.Rr),
  (ri.Ec[ri.Xr.jc] = et.Rr),
  (ri.Ec[ri.Xr.hr] = et.hr),
  (ri.Ec[ri.Xr.Ee] = et.Ee),
  (ri.Ec[ri.Xr.Cc] = et.Ee),
  (ri.Ec[ri.Xr.Bo] = et.Bo),
  (ri.Ec[ri.Xr.Bs] = et.Bs);
