// @braze/web-sdk/src/InAppMessage/models/html-message.js
import InAppMessage from "./in-app-message.js";
export default class HtmlMessage extends InAppMessage {
  constructor(i, o, e, r, d, t, s, v, n, u, a, c) {
    super(
      i,
      void 0,
      void 0,
      o,
      e,
      void 0,
      void 0,
      void 0,
      (r = r || InAppMessage.DismissType.MANUAL),
      d,
      void 0,
      void 0,
      void 0,
      void 0,
      void 0,
      void 0,
      void 0,
      void 0,
      t,
      s,
      void 0,
      void 0,
      void 0,
      v,
      void 0,
      void 0,
      void 0,
      n,
      u,
      c,
    ),
      (this.messageFields = a),
      (this.messageFields = a);
  }
  Po() {
    return !1;
  }
  $t(i) {
    if (this.uo === InAppMessage.po.Vr) {
      if (this.So) return !1;
      this.So = !0;
    }
    return this.ti.X(i), !0;
  }
  bt() {
    const i = super.bt(HtmlMessage.ds);
    return (i[InAppMessage.es.Go] = this.messageFields), i;
  }
  static Ur(i) {
    return new HtmlMessage(
      i[InAppMessage.es.ra],
      i[InAppMessage.es.vs],
      i[InAppMessage.es.ta],
      i[InAppMessage.es.pa],
      i[InAppMessage.es.ma],
      i[InAppMessage.es.ja],
      i[InAppMessage.es.xa],
      i[InAppMessage.es.wa],
      i[InAppMessage.es.ka],
      i[InAppMessage.es.CSS],
      i[InAppMessage.es.Go],
      i[InAppMessage.es.qa],
    );
  }
}
HtmlMessage.ds = InAppMessage.po.Vr;
