// @fs/zion-admin-mode/src/AdminMode.js
import React, { Suspense, lazy, useEffect } from 'react'
import PropTypes from 'prop-types'
import { useTranslation } from 'react-i18next'
import { useRoleIndicator } from '@fs/zion-role-indicator'
import { useOverlay, colors } from '@fs/zion-ui'

const ExitAdminModeOverlay = lazy(() => import('./ExitAdminModeOverlay'))

export default function AdminMode({ roleName, children }) {
  const overlay = useOverlay()
  const { handleClick } = overlay
  const [t] = useTranslation()
  const { setRole } = useRoleIndicator()

  useEffect(() => {
    setRole({
      text: roleName,
      color: colors.red70,
      onExit: handleClick,
      exitTitle: t('roles.exit-role', 'Exit Role'),
    })
  }, [setRole, roleName, handleClick, t])

  return (
    <>
      <Suspense fallback="">
        {(overlay.isOpen || overlay.transitioning) && <ExitAdminModeOverlay overlay={overlay} roleName={roleName} />}
      </Suspense>
      {children}
    </>
  )
}

AdminMode.propTypes = {
  /** roleName required */
  roleName: PropTypes.string.isRequired,
  /** Child nodes required */
  children: PropTypes.node.isRequired,
}
