// @braze/web-sdk/src/util/request-header-utils.js
import { REQUEST_ATTEMPT_DEFAULT as Ae } from "../common/constants.js";
import { STORAGE_KEYS as s } from "../managers/storage-manager.js";
const h = {
  O: {
    Nu: "data",
    gi: "content_cards/sync",
    Le: "feature_flags/sync",
    Gr: "template",
    K: "banners/sync",
  },
  Qu: (t) => (null == t ? void 0 : t.ft(s.gt.lE)),
  sm: (t) => (null == t ? void 0 : t.ft(s.gt.SE)),
  nm: (t, e) => {
    null == t || t.Bt(s.gt.lE, e);
  },
  rm: (t, e) => {
    null == t || t.Bt(s.gt.SE, e);
  },
  Bu: (t, e) => {
    if (!t || !e) return -1;
    const s = h.Qu(t);
    if (null == s) return -1;
    const n = s[e];
    return null == n || isNaN(n) ? -1 : n;
  },
  zu: (t, e) => {
    let s = Ae;
    if (!t || !e) return s;
    const n = h.sm(t);
    return null == n ? s : ((s = n[e]), null == s || isNaN(s) ? Ae : s);
  },
  W: (t, e, s) => {
    if (!t || !e) return;
    let n = h.Qu(t);
    null == n && (n = {}), (n[e] = s), h.nm(t, n);
  },
  ju: (t, e, s) => {
    if (!t || !e) return;
    let n = h.sm(t);
    null == n && (n = {}), (n[e] = s), h.rm(t, n);
  },
  wi: (t, e) => {
    t && e && h.ju(t, e, Ae);
  },
  Pu: (t, e) => {
    if (!t || !e) return;
    const s = h.zu(t, e);
    h.ju(t, e, s + 1);
  },
};
export default h;
