// @fs/zion-record-details/src/recordAnalyticsHooks.js
import { useEffect } from 'react'
import { trackEvent, usePageView } from '@fs/zion-analytics'
import zionDebug from '@fs/zion-debug'
import { useCollectionId } from './RecordProvider'

const debug = zionDebug('search:record:recordAnalyticsHooks')

/**
 * Used in Persona (in SimilarRecordInfoSheet) and Record to send a record page view event with collection ID.
 */
export default function useCollectionRecordPageView() {
  const { collectionId } = useCollectionId()
  const analyticsData = {
    event_name: 'collection_record_view',
    page_detail: 'Record Details',
    record_collectionIds: [collectionId],
  }
  usePageView(analyticsData) // merges your parameters with the defaults, then sends utag.view(pageData) to Adobe Launch
}

/**
 * For side sheets, we need to use trackEvent instead of usePageView.
 * Otherwise, the analytics page data gets confused.
 */
export function useCollectionRecordEvent() {
  const { ark, collectionId } = useCollectionId()

  // For side sheets, we need to use trackEvent instead of usePageView.
  // Otherwise, the analytics page data gets confused.
  // It should be run once per record, so we include the ark as a dependency to ensure this re-runs when the ark changes,
  // even if the collection id does not change.
  useEffect(() => {
    if (ark && collectionId) {
      const analyticsData = {
        event_name: 'collection_record_view',
        list_type: 'Record Preview',
        record_collectionIds: [collectionId],
      }
      debug('tracking record details side sheet view as an event', analyticsData)
      trackEvent(analyticsData) // merges your parameters with the defaults, then sends utag.view(pageData) to Adobe Launch
    }
  }, [ark, collectionId])
}
